PXC-3416 Fix memory leaks in garbd when started with invalid group name

venkatesh-prasad-v opened the pull request:
https://jira.percona.com/browse/PXC-3416

When garbd is started with an invalid group name, GCS didn't perform
proper cleanup before shutting down the GCS objects there by causing the
memory leaks.

This patch fixes the majority of the memory leaks when garbd is started
with invalid group name.


Valgrind report leak before the fix:
```
2020-09-21 12:04:15.847 ERROR: failed to open gcomm backend connection: 110: failed to reach primary view (pc.wait_prim_timeout): 110 (Connection timed out)
	 at gcomm/src/pc.cpp:connect():159
2020-09-21 12:04:15.848 ERROR: gcs/src/gcs_core.cpp:gcs_core_open():220: Failed to open backend connection: -110 (Connection timed out)
2020-09-21 12:04:15.875 ERROR: gcs/src/gcs.cpp:gcs_open():1700: Failed to open channel 'amy_pxc' at 'gcomm://127.0.0.1:4003': -110 (Connection timed out)
2020-09-21 12:04:15.880 FATAL: Exception in creating receive loop: Failed to open connection to group: 110 (Connection timed out)
	 at garb/garb_gcs.cpp:Gcs():35
==64814== 
==64814== HEAP SUMMARY:
==64814==     in use at exit: 1,645,275 bytes in 99 blocks
==64814==   total heap usage: 10,199 allocs, 10,100 frees, 3,251,492 bytes allocated
==64814== 
==64814== 304 bytes in 1 blocks are possibly lost in loss record 41 of 63
==64814==    at 0x483DD99: calloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==64814==    by 0x40149CA: allocate_dtv (dl-tls.c:286)
==64814==    by 0x40149CA: _dl_allocate_tls (dl-tls.c:532)
==64814==    by 0x4876322: allocate_stack (allocatestack.c:622)
==64814==    by 0x4876322: pthread_create@@GLIBC_2.2.5 (pthread_create.c:660)
==64814==    by 0x16B1B2: GCommConn::connect(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) (gcs_gcomm.cpp:277)
==64814==    by 0x163F1D: gcomm_open(gcs_backend*, char const*, bool) (gcs_gcomm.cpp:859)
==64814==    by 0x14C836: gcs_core_open(gcs_core*, char const*, char const*, bool) (gcs_core.cpp:214)
==64814==    by 0x1413F5: gcs_open(gcs_conn*, char const*, char const*, bool) (gcs.cpp:1678)
==64814==    by 0x125E35: garb::Gcs::Gcs(gu::Config&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (garb_gcs.cpp:30)
==64814==    by 0x130D00: garb::RecvLoop::RecvLoop(garb::Config const&) (garb_recv_loop.cpp:34)
==64814==    by 0x11FFEA: garb::main(int, char**) (garb_main.cpp:93)
==64814==    by 0x120273: main (garb_main.cpp:115)
==64814== 
==64814== 45,349 (832 direct, 44,517 indirect) bytes in 1 blocks are definitely lost in loss record 57 of 63
==64814==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==64814==    by 0x17D51C: gcomm::Transport::create(gcomm::Protonet&, gu::URI const&) (transport.cpp:72)
==64814==    by 0x16B2EA: GCommConn::connect(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) (gcs_gcomm.cpp:303)
==64814==    by 0x163F1D: gcomm_open(gcs_backend*, char const*, bool) (gcs_gcomm.cpp:859)
==64814==    by 0x14C836: gcs_core_open(gcs_core*, char const*, char const*, bool) (gcs_core.cpp:214)
==64814==    by 0x1413F5: gcs_open(gcs_conn*, char const*, char const*, bool) (gcs.cpp:1678)
==64814==    by 0x125E35: garb::Gcs::Gcs(gu::Config&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (garb_gcs.cpp:30)
==64814==    by 0x130D00: garb::RecvLoop::RecvLoop(garb::Config const&) (garb_recv_loop.cpp:34)
==64814==    by 0x11FFEA: garb::main(int, char**) (garb_main.cpp:93)
==64814==    by 0x120273: main (garb_main.cpp:115)
==64814== 
==64814== 1,599,622 (640 direct, 1,598,982 indirect) bytes in 1 blocks are definitely lost in loss record 63 of 63
==64814==    at 0x483DD99: calloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==64814==    by 0x13E073: gcs_create(gu_config*, gcache_t*, char const*, char const*, int, int) (gcs.cpp:310)
==64814==    by 0x125E1A: garb::Gcs::Gcs(gu::Config&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (garb_gcs.cpp:19)
==64814==    by 0x130D00: garb::RecvLoop::RecvLoop(garb::Config const&) (garb_recv_loop.cpp:34)
==64814==    by 0x11FFEA: garb::main(int, char**) (garb_main.cpp:93)
==64814==    by 0x120273: main (garb_main.cpp:115)
==64814== 
==64814== LEAK SUMMARY:
==64814==    definitely lost: 1,472 bytes in 2 blocks
==64814==    indirectly lost: 1,643,499 bytes in 96 blocks
==64814==      possibly lost: 304 bytes in 1 blocks
==64814==    still reachable: 0 bytes in 0 blocks
==64814==         suppressed: 0 bytes in 0 blocks
==64814== 
==64814== For lists of detected and suppressed errors, rerun with: -s
==64814== ERROR SUMMARY: 3 errors from 3 contexts (suppressed: 0 from 0)
```
Valgrind report leak after the fix:
```
2020-09-23 19:38:26.765 ERROR: failed to open gcomm backend connection: 110: failed to reach primary view (pc.wait_prim_timeout): 110 (Connection timed out)
	 at gcomm/src/pc.cpp:connect():159
2020-09-23 19:38:26.766 ERROR: gcs/src/gcs_core.cpp:gcs_core_open():220: Failed to open backend connection: -110 (Connection timed out)
2020-09-23 19:38:27.790  INFO: gcomm: joining thread
2020-09-23 19:38:27.810 ERROR: gcs/src/gcs.cpp:gcs_open():1700: Failed to open channel 'my_pxc1' at 'gcomm://127.0.0.1:4003': -110 (Connection timed out)
2020-09-23 19:38:27.818 FATAL: Exception in creating receive loop: Failed to open connection to group: 110 (Connection timed out)
	 at garb/garb_gcs.cpp:Gcs():35
==159096== 
==159096== HEAP SUMMARY:
==159096==     in use at exit: 0 bytes in 0 blocks
==159096==   total heap usage: 10,211 allocs, 10,211 frees, 3,253,419 bytes allocated
==159096== 
==159096== All heap blocks were freed -- no leaks are possible
==159096== 
==159096== For lists of detected and suppressed errors, rerun with: -s
==159096== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
Steps to Reproduce:
```
$ valgrind --leak-check=full $PWD/garb/garbd --donor <valid> --address "gcomm://127.0.0.1:4003" --group <invalid_group_name> --options 'base_port=12005;
```

Testing
-----
Jenkins: https://pxc.cd.percona.com/view/PXC%208.0/job/pxc-8.0-param/97/testReport/
Summary: Failing tests sporadic in nature and are unrelated to the current fix. Reported [PXC-3431](https://jira.percona.com/browse/PXC-3431) for the failing tests.

kamil-holubicki commented on gcs/src/gcs.cpp:
```diff
@@ -1762,47 +1762,46 @@ long gcs_destroy (gcs_conn_t *conn)
     gu_free (conn);
 
-    return 0;
+    return err;
```
I think swapping variables names err <-> ret would be less confusing. Now the function returns err, while we have assignments to res in the meantime.

venkatesh-prasad-v replied:
Fixed it by capturing the return values from the cleanup functions and assigning it to `err`.

kamil-holubicki commented on gcs/src/gcs.cpp:
```diff
@@ -1762,47 +1762,46 @@ long gcs_destroy (gcs_conn_t *conn)
+    /* this should cancel all recv calls */
+    gu_fifo_destroy (conn->recv_q);
+
+    if ((ret = gcs_fifo_lite_destroy (conn->repl_q)))
```
If we fail here, err is 0, so the function will return 0. Previously it returned error code.

venkatesh-prasad-v replied:
Fixed it by capturing the return values from the cleanup functions and assigning it to `err`.

kamil-holubicki commented on gcs/src/gcs.cpp:
```diff
@@ -1762,47 +1762,46 @@ long gcs_destroy (gcs_conn_t *conn)
-    if ((err = gcs_core_destroy (conn->core))) {
-        gu_debug ("Error destroying core: %d (%s)", err, strerror(-err));
-        return err;
+    if ((ret = gcs_core_close(conn->core)))
```
If we fail here, err is 0, so the function will return 0. Previously it returned error code.

venkatesh-prasad-v replied:
Fixed it by capturing the return values from the cleanup functions and assigning it to `err`.

kamil-holubicki commented on gcs/src/gcs.cpp:
```diff
@@ -1762,47 +1762,46 @@ long gcs_destroy (gcs_conn_t *conn)
+        gu_debug ("Failed to close GCS: error: %ld (%s)",-ret, strerror(-ret));
     }
 
+    if ((ret = gcs_core_destroy (conn->core)))
```
If we fail here, err is 0, so the function will return 0. Previously it returned error code.

venkatesh-prasad-v replied:
Fixed it by capturing the return values from the cleanup functions and assigning it to `err`.

kamil-holubicki replied:
As we discussed on slack:
1. Please remove ret variable as only err is enough
2. Please add the comment describing why we do not return after 1st failure and it is still worth calling gcs_core_destroy() even if gcs_core_close() failed.

venkatesh-prasad-v replied:
Done

kamil-holubicki commented on gcs/src/gcs_gcomm.cpp:
```diff
@@ -230,6 +230,18 @@ class GCommConn : public Consumer, public Toplay
+        {
+            delete tp_;
+        }
+        if (!terminated_)
```
Shouldn't terminated_ be tested under the lock Lock lock(mutex_);?
Shouldn't we signal terminated_=true and then join?

venkatesh-prasad-v replied:
Fixed this by calling `notify()` after setting `terminated_ = true` with mutex protection.

venkatesh-prasad-v commented:
Jenkins: https://pxc.cd.percona.com/view/PXC%208.0/job/pxc-8.0-param/101/testReport/ [In progress]

venkatesh-prasad-v commented:
The test failure of  `galera.galera_restart_on_unknown_option` seems to be related to the current fix and this failure needs to be investigated

venkatesh-prasad-v commented:
Fixed the test failure of `galera.galera_restart_on_unknown_option`.
Jenkins: https://pxc.cd.percona.com/view/PXC%208.0/job/pxc-8.0-param/104/testReport/

Commit:
PXC-3416 Fix memory leaks in garbd when started with invalid group name

https://jira.percona.com/browse/PXC-3416

When garbd is started with an invalid group name, GCS didn't perform
proper cleanup before shutting down the GCS objects there by causing the
memory leaks.

This patch fixes the majority of the memory leaks when garbd is started
with invalid group name.

venkatesh-prasad-v commented:
Jenkins: https://pxc.cd.percona.com/view/PXC%208.0/job/pxc-8.0-param/106/testReport/ 

Summary: Failing tests are unrelated to the fix.

kamil-holubicki commented on gcs/src/gcs_gcomm.cpp:
```diff
@@ -230,6 +230,29 @@ class GCommConn : public Consumer, public Toplay
+            gu_thread_join(thd_, 0);
+            mutex_.lock();
+        }
+        mutex_.unlock();
```
LGTM
I don't know if it would be better, probably yes because of getting rid of unnecessary lock/unlock cycle, but consider this setting unlock_needed = true before 'if', then unlock_needed = false in 'if' after unlocking and then if(unlock_needed) mutex._unlock() after 'if'

venkatesh-prasad-v replied:
Done

Commit:
Address review comments from Kamil

kamil-holubicki reviewed (approved):
lgtm
